ls1intum/Artemis
# `Plagiarism checks`: Mark programming submissions detected by cpc as plagiarism and reduce their score

# ⚠️ This PR contains braking db changes. Please use ts4 to deploy it. More notes in _Steps for Testing_

### Checklist

#### General


#### Server
#### Client
#### Changes affecting Programming Exercises
- [x] I tested **all** changes and their related features with **all** corresponding user types on Test Server 4 (due to breaking db changes)
### Motivation and Context
My previous PR #6666 introduced cpc, but until this PR there was no immediate feedback delivered to the student about plagiarism detected in his or her submission. Changes introduces here notify the student about suspected plagiarism.

Due to an issue with the current implementation of handling results of plagiarism checks for programming exercises this PR contains also changes in how data is exchanged between Jplag and Artemis.

This implements CPC-3 and CPC-5 milestone. More about that can be discovered in my thesis proposal: https://confluence.ase.in.tum.de/display/ArTEMiS/MA+Jakub+Riegel?preview=/157431627/166332701/riegel_proposal.pdf

### Description
This PR adds programming exercises to the features introduced in #6804. Please check its description for detailed information.

### Steps for Testing
- 1 Instructor
- 2 Students
- 1 Course
- for convenience use the `cpc-mark-submissions-as-plagiarism-programing-1min` [branch](https://github.com/ls1intum/Artemis/tree/cpc-mark-submissions-as-plagiarism-programing-1min) which triggers cpc every 1 minute
- **due to database changes please deploy it to ts4 by using #6946** 

General:
1. Create programming exercise with continuous plagiarism control enabled (use _Enabled_ checkbox on exercise creation form)
4. Create two similar submission for the exercise
5. Wait 2 minutes (results need to synchronise)
6. Verify that ⚠️ _Suspicion of plagiarism_ warning was placed and 0 points result was posted (see screenshots at the bottom)
7. Change one submission to something completely different
8. Wait 2 minutes
9. Verify that warnings and results are deleted

Disputing results:
1. Generate submissions with plagiarism detected by the cpc (using the steps above)
2. Change exercise due date to the past
3. As a student go into your submission page and dispute the result
4. As instructor approve and decline the dispute
5. If approved: new scores are displayed; if not: old results are displayed 

### Review Progress

#### Performance Review
#### Code Review
#### Manual Tests
#### Exam Mode Test

### Test Coverage
| Class/File | Line Coverage | Confirmation (assert/expect) | Note |
|------------|--------------:|-----------------------------:|:---:|
| TextPlagiarismResultConverter.java | 54% | ✅                           | Logic of `TextPlagiarismResultConverter` was implemented already and this PR just moves it to separate class. Those methods were not tested before. It is out of the scope of cpc to implement detailed tests for them. There is an issue to do it: .#6966  |
| JPlagSubmissionDataExtractor.java | 86% | ✅                           | - |

### Screenshots
Programming exercise page header when cpc detects plagiarism:
<img width="1512" alt="image" src="https://github.com/ls1intum/Artemis/assets/32958017/a7800f42-2fcd-4a55-881d-12e9ce6c6e6a">
